### PR TITLE
 Allocations tests for SYMMLQ

### LINKED ISSUE #97: Allocations tests
We should check the allocations of each solver (old and future ones). It's a good way to find mistakes and verify that implementations perform well.
# Allocations tests roadmap
## Methods already in Krylov.jl
- [x] CG_LANCZOS
- [x] CG_LANCZOS_SHIFT_SEQ
- [x] CG
- [x] CR
- [x] CGLS
- [x] CRLS
- [x] CGNE
- [x] CRAIG
- [x] CRMR
- [x] CRAIGMR
- [x] LSLQ
- [x] LSMR
- [x] LSQR
- [x] MINRES
- [x] SYMMLQ
- [x] DQGMRES
- [x] DIOM
- [x] CGS
## Future methods
- [x] USYMLQ
- [x] USYMQR
- [x] TFQMR
- [x] GMRES
- [x] FOM

### PR DESCRIPTION
**Pull request by amontoison**
fix #97

### CODE REVIEW

File: src/symmlq.jl [hunk 1]
```diff
@@ -70,14 +74,11 @@ function symmlq(A :: AbstractLinearOperator, b :: AbstractVector{T};
   cold = 1.0
   sold = 0.0
 
-  wbar = vold
   ζold = 0.0
   ζbar = β₁/γbar
 
   ANorm² = α * α + β * β
 
-  xcg = ζbar*vold
```
Comment (dpo) [inline src/symmlq.jl]: I can't find where this line went.

File: src/symmlq.jl [hunk 2]
```diff
@@ -260,9 +262,12 @@ function symmlq(A :: AbstractLinearOperator, b :: AbstractVector{T};
     solved = solved_mach | solved_lim | zero_resid_mach | zero_resid_lim | fwd_err
   end
 
+  # Compute CG point
+  @. x_cg = x_lq + ζbar * wbar
```
Comment (dpo) [inline src/symmlq.jl]: Do you only update x_cg once at the end?

Comment (amontoison) [reply src/symmlq.jl]: Yes

File: src/symmlq.jl (hunk 1 above)
Comment (amontoison) [reply src/symmlq.jl]: I removed it. We can update `x_cg` only one time  at the end even if `iter=1`.

**Comment by dpo** [comment]
Ok, thank you. Technically, we could do without x_cg. It's really there for information. We could update x_lq one last time and save a vector. As an intermediate measure, we could overwrite wbar with x_cg.

**Comment by amontoison** [comment]
Maybe it's possible to use one less vector in the Lanczos process (`MINRES`, `CG_LANCZOS`, `SYMMLQ` methods)

Currently our implementation is :
one preallocated vector Av for Avₖ, one vector for vₖ and one vector for vₖ₋₁ for a total of three vectors.
Then we update Av with some @kaxpy! to have this relation : 
`βₖ₊vₖ₊₁ = Avₖ - δₖvₖ - βₖvₖ₋₁` with `δₖ = vₖᵀ Avₖ`

But I think we can build `vₖ₊₁` from `vₖ₋₁`, It could be possible to do a @kscal! to compute βₖvₖ₋₁
and then compute βₖvₖ₋₁ + Avₖ directly. Like this for example `vₖ₋₁[i] = βₖvₖ₋₁[i] * sum(A[i,j] * vₖ[j] for j=1:n)` (basic form without BLAS operation).

It's still possible to compute δₖ because vₖᵀ ( βₖvₖ₋₁ + Avₖ) =  vₖᵀ Avₖ and vₖ₋₁ ⟂ vₖ.
We finish to update vₖ₊₁ with `Avₖ - βₖvₖ₋₁ - δₖvₖ` and after we do a `@kswap` to swich pointers.

We can use only two vectors with that, what do you think?

**Comment by amontoison** [comment]
I like the idea to overwrite `wbar` with `x_cg`.
```julia
@kaxpy!(n, 1.0, x_lq, ζbar, wbar)
x_cg = wbar
```
You already merge but we can do it in the future.

**Comment by dpo** [comment]
Feel free to open a new PR. It will be simple, and we won't have to give up x_cg. The storage is already there!

**Comment by amontoison** [comment]
Yes, you're right. I can do now, it's only two lines to write.

**Pull request merged**